Sources/SwiftFormat/Utilities/FileIterator.swift

```diff
@@ -173,9 +159,34 @@ public struct FileIterator: Sequence, IteratorProtocol {
   }
 }
 
-/// Returns the type of the file at the given URL.
-private func fileType(at url: URL) -> FileAttributeType? {
-  // We cannot use `URL.resourceValues(forKeys:)` here because it appears to behave incorrectly on
-  // Linux.
-  return try? FileManager.default.attributesOfItem(atPath: url.path)[.type] as? FileAttributeType
+/// Returns the actual URL and type of the file at the given URL, following symlinks if requested.
+///
+/// - Parameters:
+///   - url: The URL to get the file and type of.
+///   - followSymlinks: Whether to follow symlinks.
+/// - Returns: The actual URL and type of the file at the given URL, or `nil` if the file does not
+///   exist or is not a supported file type. If `followSymlinks` is `true`, the returned URL may be
+///   different from the given URL; otherwise, it will be the same.
+private func fileAndType(at url: URL, followSymlinks: Bool) -> (URL, FileAttributeType)? {
+  func typeOfFile(at url: URL) -> FileAttributeType? {
+    // We cannot use `URL.resourceValues(forKeys:)` here because it appears to behave incorrectly on
+    // Linux.
+    return try? FileManager.default.attributesOfItem(atPath: url.path)[.type] as? FileAttributeType
+  }
+
+  guard var fileType = typeOfFile(at: url) else {
+    return nil
+  }
+
+  var url = url
+  while followSymlinks && fileType == .typeSymbolicLink,
+    let destination = try? FileManager.default.destinationOfSymbolicLink(atPath: url.path)
```

Comment:
Should we detect circles here if you have a symlink that points to itself?

Reply:
Good catch; fixed. Symlink cycles (either a link pointing to itself, or a multi-link cycle) are silently ignored now. We could warn on this, but we'd have to plumb `DiagnosticsEngine` down into `FileIterator`, which I didn't really want to do here if we can avoid it.

Reply:
As long as we don’t infinite loop, I think the behavior is OK.

Also, it only just occurs to me, why can’t we use `resolvingSymlinksInPath`/`realpath` here if we follow all symlinks?

In SourceKit-LSP, we have the following code to avoid the `/tmp` vs `/private/tmp` issue: https://github.com/swiftlang/sourcekit-lsp/blob/3e7f350246dc86072a9dfc88f25782dc86dd1c9a/Sources/SwiftExtensions/URLExtensions.swift#L40-L57

Reply:
Hmm, good question. That `realpath` code seems to work on macOS from local testing, so let me see if it works on Linux/Windows. Assuming they both consistently resolve all symlinks deeply or stop when a cycle is reached on all platforms, then that would remove the need for our own loop.

Reply:
Sigh, no good. Based on CI results and local testing, it looks like `resolvingSymlinksInPath` on Linux pre-5.x Foundation doesn't correctly resolve through multiple symlinks. So we'd need a special case to handle those anyway, which would essentially be the code I had earlier anyway.

Maybe one day when we drop 5.x support, we can clean this all up.